src/Workspaces/Core/Portable/Remote/RemoteHostSessions.cs

```diff
@@ -0,0 +1,280 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Execution;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Remote
+{
+    /// <summary>
+    /// Helper type of common session case
+    /// 
+    /// RemoteHostClient.Session with 1 solution
+    /// </summary>
+    internal sealed class SessionWithSolution : IDisposable
+    {
+        private readonly RemoteHostClient.Session _session;
+        private readonly PinnedRemotableDataScope _scope;
+        private readonly CancellationToken _cancellationToken;
+
+        public static async Task<SessionWithSolution> CreateAsync(RemoteHostClient.Session session, PinnedRemotableDataScope scope, CancellationToken cancellationToken)
+        {
+            var sessionWithSolution = new SessionWithSolution(session, scope, cancellationToken);
+
+            try
+            {
+                await session.RegisterPinnedRemotableDataScopeAsync(scope).ConfigureAwait(false);
+                return sessionWithSolution;
+            }
+            catch
+            {
+                sessionWithSolution.Dispose();
+                return null;
+            }
+        }
+
+        private SessionWithSolution(RemoteHostClient.Session session, PinnedRemotableDataScope scope, CancellationToken cancellationToken)
+        {
+            _session = session;
+            _scope = scope;
+            _cancellationToken = cancellationToken;
+        }
+
+        public void AddAdditionalAssets(CustomAsset asset)
+        {
+            _scope.AddAdditionalAsset(asset, _cancellationToken);
+        }
+
+        public void Dispose()
+        {
+            _scope.Dispose();
+            _session.Dispose();
+        }
+
+        public Task InvokeAsync(string targetName, params object[] arguments) =>
+            _session.InvokeAsync(targetName, arguments);
+        public Task<T> InvokeAsync<T>(string targetName, params object[] arguments) =>
+            _session.InvokeAsync<T>(targetName, arguments);
+        public Task InvokeAsync(string targetName, IEnumerable<object> arguments, Func<Stream, CancellationToken, Task> funcWithDirectStreamAsync) =>
+            _session.InvokeAsync(targetName, arguments, funcWithDirectStreamAsync);
+        public Task<T> InvokeAsync<T>(string targetName, IEnumerable<object> arguments, Func<Stream, CancellationToken, Task<T>> funcWithDirectStreamAsync) =>
+            _session.InvokeAsync<T>(targetName, arguments, funcWithDirectStreamAsync);
+    }
+
+    /// <summary>
+    /// Helper type of common session case
+    /// 
+    /// RemoteHostClient.Session that lives for multiple solutions
+    /// </summary>
+    internal sealed class KeepAliveSession
```

Comment:
since this is not supposed to be used in using. it has Shutdown instead. I probably just remove name session to remove confusion.